File: javascript/src/socketd/Socketd.ts
/**
 * let connentor = new ClientConnector(this.config);
 *  return new Session(new ClientChannel(connentor.connect(), connentor));
 * */
import {Client} from "./transport/client/Client";
import {ClientConfig} from "./transport/client/ClientConfig";
import {Entity} from "./transport/core/Entity";
import {WsClient} from "./impl_ws/WsClient";


export const SocketD = {
    createClient(url): Client {
        return new WsClient(new ClientConfig(url));
    }
}

File: go/socketd/socketd.go
package socketd

import (
	"fmt"

	"socketd/transport/client"

	"socketd/transport/server"
	"socketd/transprot-impl/tcp"
)

var SocketD = new(socketD)

type socketD struct {
	cfg *Config
}

func (sd *socketD) Version() string {
	return "0.1.0"
}

func (sd *socketD) ProtocolVersion() string {
	return "1.0"
}

func (sd *socketD) Config(options ...ConfigOption) *socketD {
	sd.cfg = new(Config)
	for _, opts := range options {
		opts(sd.cfg)
	}
	return sd
}

func (sd *socketD) CreateServer() server.Server {
	if sd.cfg == nil {
		panic("use Config(…) before creating")
	}
	switch sd.cfg.GetSchema() {
	case "tcp":
		return tcp.NewTcpServer(sd.cfg)
	}
	panic(fmt.Errorf("%s server not implement", sd.cfg.schema))
}

func (sd *socketD) CreateClient() client.Client {
	if sd.cfg == nil {
		panic("use Config(…) before creating")
	}
	switch sd.cfg.GetSchema() {
	case "tcp":
		return tcp.NewTcpClient(sd.cfg)
	}
	panic(fmt.Errorf("%s server not implement", sd.cfg.schema))
}
